### PR TITLE
feature/san-chart

### PR DESCRIPTION
**Pull request by DmitriVanGuard**
## Summary
Replacing recharts with the in-house chart library

## Supported features
- Multicharts;
- Wide chart;
- Paywall;
- Watermark;
- Saving chart as PNG;
- Referencing anomalies and trends;
- Last day price reference line;
- Social context;
- Histogram;
- Signals (if logged in && beta user && has price metric selected);
  - Reference line with a description;
  - Viewing existing signal description and reference line by hovering on it;
  - Signal creation by clicking on the reference line start;
  - Deleting signal by clicking on it.

## Screenshots
gif

### CODE REVIEW

File: src/ducks/SANCharts/Chart/axes.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import { getDateDayMonthYear } from './utils'
+
+const axesTickFormatters = {
+  datetime: getDateDayMonthYear
+}
+
+export function plotAxes (chart) {
+  const { ctx } = chart
+  drawAxes(chart)
+  ctx.fillStyle = '#9faac4'
```
Comment (Partysun) [inline src/ducks/SANCharts/Chart/axes.js]: Is it a color from our design system? @DmitriVanGuard

File: src/ducks/SANCharts/Chart/index.js [hunk 2]
```diff
@@ -151,6 +143,10 @@ const Chart = ({
     events.forEach(({ metric, key, datetime, value, color }) =>
       drawReferenceDot(chart, metric, datetime, color, key, value)
     )
+
+    /* if (!hasPremium) { */
```
Comment (Partysun) [inline src/ducks/SANCharts/Chart/index.js]: @DmitriVanGuard Why do we have this comments?

File: src/pages/Chart/paywallBoundaries.js [hunk 3]
```diff
@@ -1,40 +1,59 @@
 import { getTimeIntervalFromToday, DAY, MONTH } from '../../utils/dates'
 
-const options = { isUTC: true }
+export function createOptions (hours = 48) {
```
Comment (Partysun) [inline src/pages/Chart/paywallBoundaries.js]: @DmitriVanGuard Why do we need to export this func? For tests?

File: src/pages/Chart/paywallBoundaries.js [hunk 4]
```diff
@@ -1,40 +1,59 @@
...
-  options
+  createOptions()
 )
 const PRO_RIGHT_BOUNDARY_DATE = false
 
 const ENTERPRISE_LEFT_BOUNDARY_DATE = false
 const ENTERPRISE_RIGHT_BOUNDARY_DATE = false
 
 export default {
+  ANON: {
```
Comment (Partysun) [inline src/pages/Chart/paywallBoundaries.js]: @DmitriVanGuard Why do we need to precalculate boundaries for anon and rest (all) plans? I dont' check full file, see only a part.

File: src/ducks/SANCharts/Chart/paywall.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  ctx.setLineDash([7])
+
+  if (lastDate <= leftBoundaryDate || firstDate >= rightBoundaryDate) {
+    ctx.strokeStyle = 'red'
+    ctx.strokeRect(left, top, width, height)
+    ctx.restore()
+    return
+  }
+
+  const factor = lastIndex / (lastDate - firstDate)
```
Comment (Partysun) [inline src/ducks/SANCharts/Chart/paywall.js]: @DmitriVanGuard is it possible, where points.length = 1? Cause in that case we can have an error here.

File: src/ducks/SANCharts/Chart/paywall.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+export function drawPaywall (chart, leftBoundaryDate, rightBoundaryDate) {
+  const { ctx, points, left, top, width, height } = chart
+
+  const lastIndex = points.length - 1
+  const firstDate = points[0].value
+  const lastDate = points[lastIndex].value
+
+  ctx.save()
+  ctx.lineWidth = 1
+  ctx.strokeStyle = '#FFAD4D'
```
Comment (Partysun) [inline src/ducks/SANCharts/Chart/paywall.js]: @DmitriVanGuard Do we have this color in our design system?

File: src/ducks/SANCharts/Chart/axes.js (hunk 1 above)
Comment (DmitriVanGuard) [reply src/ducks/SANCharts/Chart/axes.js]: https://storybook.stage.internal.santiment.net/?selectedKind=Colors&selectedStory=List%20of%20all%20colors&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel

File: src/pages/Chart/paywallBoundaries.js (hunk 4 above)
Comment (DmitriVanGuard) [reply src/pages/Chart/paywallBoundaries.js]: What is the alternative?

File: src/ducks/SANCharts/Chart/paywall.js (hunk 6 above)
Comment (DmitriVanGuard) [reply src/ducks/SANCharts/Chart/paywall.js]: https://storybook.stage.internal.santiment.net/?selectedKind=Colors&selectedStory=List%20of%20all%20colors&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel

File: src/ducks/SANCharts/Chart/paywall.js (hunk 5 above)
Comment (DmitriVanGuard) [reply src/ducks/SANCharts/Chart/paywall.js]: What is the reason for error?

Comment (Partysun) [reply src/ducks/SANCharts/Chart/paywall.js]: If lastDate will equal firstData

File: src/ducks/SANCharts/Chart/paywall.js (hunk 6 above)
Comment (Partysun) [reply src/ducks/SANCharts/Chart/paywall.js]: @DmitriVanGuard Before you wrote `ctx.fillStyle` = '#9faac4'` and here you are using capitalize letters

Comment (DmitriVanGuard) [reply src/ducks/SANCharts/Chart/paywall.js]: `ctx.fillStyle` is not case sensitive

File: src/ducks/SANCharts/Chart/paywall.js (hunk 5 above)
Comment (DmitriVanGuard) [reply src/ducks/SANCharts/Chart/paywall.js]: Why it should throw an error?

**Comment by DmitriVanGuard** [comment]
@Partysun @GarageInc @haritonasty please, check `Supported features` list

File: src/ducks/SANCharts/Chart/index.js [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    if (!hasPremium) {
+      drawPaywall(chart, leftBoundaryDate, rightBoundaryDate)
+    }
+  }
+
+  function marker (ctx, key, value, x, y) {
+    const RADIUS = 4
+    if (key === 'isAnomaly') {
+      ctx.beginPath()
+      ctx.arc(x + RADIUS, y + 1, RADIUS, 0, 2 * Math.PI)
```
Comment (GarageInc) [inline src/ducks/SANCharts/Chart/index.js]: Dark Magic

Comment (GarageInc) [reply src/ducks/SANCharts/Chart/index.js]: dublicated lines in some places

File: src/ducks/SANCharts/Chart/settings.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  }
+}
+
+Object.keys(Metrics).forEach(metric => {
+  const { key, dataKey = key, formatter = FROMATTER, label } = Metrics[metric]
+  tooltipSettings[dataKey] = {
+    label,
+    formatter
+  }
+})
```
Comment (haritonasty) [inline src/ducks/SANCharts/Chart/settings.js]: you can use `Object.values` for this case

File: src/ducks/SANCharts/Chart/settings.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        case 2:
+          return '2nd'
+        case 3:
+          return '3rd'
+        default:
+          return `${val}th`
+      }
+    }
+  }
+}
```
Comment (haritonasty) [inline src/ducks/SANCharts/Chart/settings.js]: Could we keep this template in 1 place? It's repeating in `data.js`

File: src/ducks/SANCharts/Chart/paywall.js (hunk 5 above)
Comment (GarageInc) [reply src/ducks/SANCharts/Chart/paywall.js]: lastIndex / 0

File: src/ducks/SANCharts/Chart/index.js [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+  return (
+    <div className={styles.wrapper}>
+      <canvas ref={canvasRef} />
+      {hasPriceMetric && (
+        <Signals chart={chart} data={data} slug={slug} scale={scale} />
+      )}
+      {isLoading && (
+        <div className={styles.loader}>
+          <Loader />
+        </div>
```
Comment (haritonasty) [inline src/ducks/SANCharts/Chart/index.js]: set `className` directly to Loader, without `div`

File: src/ducks/SANCharts/Chart/utils.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  (metrics.includes(Metrics.historyPrice) && Metrics.historyPrice) ||
+  metrics.find(({ node }) => node === 'line') ||
+  metrics[0]
+
+export function findPointIndexByDate (points, target) {
+  const lastIndex = points.length - 1
+  const firstDate = points[0].value
+  const lastDate = points[lastIndex].value
+
+  const factor = lastIndex / (lastDate - firstDate)
```
Comment (haritonasty) [inline src/ducks/SANCharts/Chart/utils.js]: this repeats in `paywall.js`. Maybe move it into separate function?

File: src/pages/Chart/paywallBoundaries.js [hunk 12]
```diff
@@ -1,40 +1,59 @@
...
+  const NEXT_DAY = new Date()
+  NEXT_DAY.setHours(hours, 0, 0, 0)
+  return { isUTC: true, from: NEXT_DAY }
+}
+
+const { from: ANON_LEFT_BOUNDARY_DATE } = getTimeIntervalFromToday(
+  -3,
+  MONTH,
+  createOptions(72)
+)
```
Comment (haritonasty) [inline src/pages/Chart/paywallBoundaries.js]: Not clear for me...you set here `72` hours? Or what's idea?

**Comment by haritonasty** [comment]
In future update don't forget to change tooltip and ticks styles and them in night mode, please)

File: src/ducks/SANCharts/Chart/Signals/helpers.js [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+const HEIGHT = 22
+
+const MARGIN_BOTTOM = 5
+const BUBBLE_MARGIN_BOTTOM = HEIGHT + MARGIN_BOTTOM
+
+const PADDING_LEFT = 8
+const PADDINGS = PADDING_LEFT * 2
+
+const MARGIN_LEFT = 5
+const TEXT_LEFT_MARGIN = MARGIN_LEFT + PADDING_LEFT
```
Comment (Partysun) [inline src/ducks/SANCharts/Chart/Signals/helpers.js]: @DmitriVanGuard Can you write which text (ticks/labels/tooltips)

**Comment by haritonasty** [comment]
- Suppose we have a small incorrection when mapping anomalies 
- When we have `0` not sure, that it's always is `no data` label
- Different boundaries for anon on detailed page and index page

File: src/ducks/SANCharts/Chart/Signals/helpers.js [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+  if (value > max || value < min) {
+    return
+  }
+
+  return {
+    id,
+    value,
+    type: below ? SIGNAL_BELOW : SIGNAL_ABOVE,
+    y: scale(height, min, max)(value) + top
```
Comment (Partysun) [inline src/ducks/SANCharts/Chart/Signals/helpers.js]: @DmitriVanGuard it's not easy to read quick. Maybe make a middle func

File: src/ducks/SANCharts/Chart/Signals/index.js [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+  fetchSignals: () => dispatch(fetchSignals()),
+  removeSignal: id => dispatch(removeTrigger(id))
+})
+
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps
+)(props => {
+  const { chart, isLoggedIn, isBeta } = props
+  return isLoggedIn && isBeta && chart ? <Signals {...props} /> : null
```
Comment (Partysun) [inline src/ducks/SANCharts/Chart/Signals/index.js]: @DmitriVanGuard We can disable beta here. If it works)

File: src/pages/Chart/paywallBoundaries.js [hunk 16]
```diff
@@ -1,40 +1,59 @@
...
+const { from: ANON_RIGHT_BOUNDARY_DATE } = getTimeIntervalFromToday(
+  -1,
+  DAY,
+  createOptions(24)
+)
```
[the remark this answers is not in the document]
Comment (DmitriVanGuard) [reply src/pages/Chart/paywallBoundaries.js]: That's how the backend deals with it

File: src/pages/Chart/paywallBoundaries.js (hunk 12 above)
Comment (Partysun) [reply src/pages/Chart/paywallBoundaries.js]: Maybe make an interface as 
getTimeIntervalFromToday({cutMoreThanLastN: '3m', cutLastN: '2d' or '24h'})

File: src/pages/Chart/paywallBoundaries.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (haritonasty) [reply src/pages/Chart/paywallBoundaries.js]: @Partysun ? It's strange. In this case user prefer use anon mode instead free/basic plan

[the remark this answers is not in the document]
Comment (Partysun) [reply src/pages/Chart/paywallBoundaries.js]: Agree. Will check it with backend and last sheet of our plans

File: src/ducks/SANCharts/Chart/paywall.js (hunk 5 above)
Comment (DmitriVanGuard) [reply src/ducks/SANCharts/Chart/paywall.js]: Ah, ok. `Infinity` as the array index later in the code

File: src/ducks/SANCharts/Chart/Signals/Signal.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
+import React from 'react'
+import styles from './index.module.scss'
+
+const Signal = ({ signal, setHovered, removeSignal }) => {
```
[the remark this answers is not in the document]
Comment (DmitriVanGuard) [reply src/ducks/SANCharts/Chart/Signals/Signal.js]: I don't know much about signals internal design. Speaking about object creation - I have copied it from the previous implementation
https://github.com/santiment/app/pull/938/files?file-filters%5B%5D=.js&file-filters%5B%5D=.lock&file-filters%5B%5D=.scss#diff-4854c1da05a0245d4a48c63c59b84c17L138

File: src/ducks/SANCharts/Chart/Signals/helpers.js (hunk 13 above)
Comment (DmitriVanGuard) [reply src/ducks/SANCharts/Chart/Signals/helpers.js]: It's for the text inside the rect: `Click to ....`
image

**Pull request merged**